### ISSUE TITLE
Add support for 2FA (OTP, U2F)

### ISSUE THREAD

**Issue by Perflyst** [omarroth/invidious#418]
Please add support for OTP (HOTP) and/or U2F. I am sure there are some librarys for this.

I personally prefer U2F as second factor more than OTP.

**Comment by omarroth** [comment]
I assume this is for Invidious login, could you provide any more details on implementing them or what apps/devices they should be tested with?

**Comment by Perflyst** [comment]
Well OTP is always probably the same. You can use a Nitrokey, Yubikey or also command line tools.
For U2F I personally can test with a Nitrokey.

**Comment by trymeouteh** [comment]
+1

This will increase account security. I always like services that support OTP for accounts. If invidious gets hacked the passwords get leaked accounts are still secure due to 2FA.

**Comment by aelisya** [comment]
i want to ask if there is any news on this feature ?
(OTP or U2F (FIDO))

**Comment by aelisya** [comment]
on the who, there is a library (MIT Licence) : https://github.com/lelag/otphp in php for doing OTP on login

**Comment by syeopite** [comment]
Unfortunately Invidious is programmed in Crystal so a php library won't be compatible. Fortunately there are libraries for this and I'll be working on implementing it very soon!

**Comment by unixfox** [comment]
> Unfortunately Invidious is programmed in Crystal so a php library won't be compatible. Fortunately there are libraries for this and I'll be working on implementing it very soon!

There is a library for crystal here: https://github.com/philnash/crotp

**Comment by syeopite** [comment]
I'm aware. I'm currently just waiting for the maintainer to officially support Crystal 1.0 but I'll go ahead and fork it if needed.

**Comment by BirdInFire** [comment]
any news on this ?

**Comment by unixfox** [comment]
No, a PR https://github.com/iv-org/invidious/pull/2254 is pending but we still don't have the time yet for finishing it.

But if you want to contribute to the pull request you are free to do so :)!

**Comment by SamantazFox** [comment]
The PR is mostly done, but we first need a database migration framework.

**Comment by BirdInFire** [comment]
> No, a PR #2254 is pending but we still don't have the time yet for finishing it.
> 
> But if you want to contribute to the pull request you are free to do so :)!

i'm open to help but i do not know anything on crystal, so i can't help for that, will simply wait just wanted to know if we can hope to see it soon for up to date instance or if i must continue without it for now).

> The PR is mostly done, but we first need a database migration framework.

Thanks for the answer ;)

**Comment by BirdInFire** [comment]
> No, a PR #2254 is pending but we still don't have the time yet for finishing it.
> 
> But if you want to contribute to the pull request you are free to do so :)!

any news ?

**Comment by unixfox** [comment]
> > No, a PR #2254 is pending but we still don't have the time yet for finishing it.
> > But if you want to contribute to the pull request you are free to do so :)!
> 
> any news ?

No, the developer that worked on this feature, [syeopite](https://github.com/syeopite), left the project so nobody is working on this feature anymore.

**Comment by BirdInFire** [comment]
> > > No, a PR #2254 is pending but we still don't have the time yet for finishing it.
> > > But if you want to contribute to the pull request you are free to do so :)!
> > 
> > 
> > any news ?
> 
> No, the developer that worked on this feature, [syeopite](https://github.com/syeopite), left the project so nobody is working on this feature anymore.

quite saf
